Flexible expectation model

seabbs opened the pull request:
This  PR closes #5 by adding a flexible and user-definable expectation model. At the same time it recasts the underlying model as the most basic possible renewal model (i.e a fixed generation time with no observation model and no between-group interactions) with the default of `generation_time = 1` resulting in a daily growth rate model. 

The decision to remove the`order` argument was taken based on the fact that keeping this significantly increases model complexity. 

Initial testing on the following is promising but does suggest quite a performance hit when the previous model (which can now be defined as `~ (1 | day)` is used (note this is hard to define on a group wise basis)).

*Note: I have done very little checking of this yet so bugs very likely. As this is quite a big feature change thought surfacing it early would be useful.*

```r
# Load packages
library(epinowcast)
library(data.table)
library(ggplot2)

# Use 2 cores
options(mc.cores = 2)

# Load and filter germany hospitalisations
nat_germany_hosp <- germany_covid19_hosp[location == "DE"][age_group %in% "00+"]
nat_germany_hosp <- enw_filter_report_dates(
  nat_germany_hosp,
  latest_date = "2021-10-01"
)

# Make sure observations are complete
nat_germany_hosp <- enw_complete_dates(
  nat_germany_hosp,
  by = c("location", "age_group")
)
# Make a retrospective dataset
retro_nat_germany <- enw_filter_report_dates(
  nat_germany_hosp,
  remove_days = 40
)
retro_nat_germany <- enw_filter_reference_dates(
  retro_nat_germany,
  include_days = 40
)

# Get latest observations for the same time period
latest_obs <- enw_latest_data(nat_germany_hosp)
latest_obs <- enw_filter_reference_dates(
  latest_obs,
  remove_days = 40, include_days = 20
)

# Preprocess observations (note this maximum delay is likely too short)
pobs <- enw_preprocess_data(retro_nat_germany, max_delay = 20)

# Fit the default nowcast model and produce a nowcast
# Note that we have reduced samples for this example to reduce runtimes
nowcast <- epinowcast(pobs,
  expectation = enw_expectation(~ rw(week) + (1 | day_of_week), data = pobs),
  fit = enw_fit_opts(
    save_warmup = FALSE, pp = TRUE,
    chains = 2, iter_warmup = 500, iter_sampling = 500
  ),
  obs = enw_obs(family = "poisson", data = pobs)
)
```

seabbs commented:
Thinking about this part of the performance hit may be because the `~ 1 + (1 | day)` variant of the model has an extra parameter (the intercept) Vs the original. Perhaps could specify the intercept as optional here. That would take some model adaptations and could potentially be a little dangerous.

seabbs commented:
I think it might be a good idea to merge this into `develop` in a very MVP state if possible to help with exploration and help make smaller PRs fleshing elements out easier. What do people think? It's all getting a bit monolithic for my tastes.

I think the minimum that needs to be done is:

- [ ] Merge in #147 
- [ ] Update this PR based on `develop`
- [ ] Decide on a default model (I think for now this should be a weekly random walk and no observation level random walk). 
- [ ] Promote to issues some of the tasks outlined above as needed (including updating the model definition)
- [ ] Add a note to the news
- [ ] Update at least the README + put a note on the case study saying currently out of date with the `develop` branch.

seabbs commented:
#154 added latent reporting to this PR (i.e observations are based on some discrete scaled convolution of an underlying latent parameter) and a new formula interface for specifying observation level effects (i.e the growth rate is constant but reported observations have a day-of-the-week adjustment. 

This in effect integrates `EpiNow2` (in its most basic form with fixed delays) into the `epinowcast` framework with the addition of formula interface flexibility. No changes have been made to make nowcasting optional but this should be implemented in the future as it increases overlap with `EpiNow2` making it easier to depreciate. 

*Note: This is a highly experimental feature and subject to change.*

```r
# Load packages
library(epinowcast)
library(data.table)
library(ggplot2)

# Use 2 cores
options(mc.cores = 2)

# Load and filter germany hospitalisations
nat_germany_hosp <- germany_covid19_hosp[location == "DE"][age_group %in% "00+"]
nat_germany_hosp <- enw_filter_report_dates(
  nat_germany_hosp,
  latest_date = "2021-10-01"
)

# Make sure observations are complete
nat_germany_hosp <- enw_complete_dates(
  nat_germany_hosp,
  by = c("location", "age_group")
)
# Make a retrospective dataset
retro_nat_germany <- enw_filter_report_dates(
  nat_germany_hosp,
  remove_days = 40
)
retro_nat_germany <- enw_filter_reference_dates(
  retro_nat_germany,
  include_days = 40
)

# Get latest observations for the same time period
latest_obs <- enw_latest_data(nat_germany_hosp)
latest_obs <- enw_filter_reference_dates(
  latest_obs,
  remove_days = 40, include_days = 20
)

# Preprocess observations (note this maximum delay is likely too short)
pobs <- enw_preprocess_data(retro_nat_germany, max_delay = 20)

# Fit the default nowcast model and produce a nowcast
# Note that we have reduced samples for this example to reduce runtimes
nowcast <- epinowcast(pobs,
  expectation = enw_expectation(
    ~ rw(week), generation_time = c(0.2, 0.6, 0.2),
    observation = ~ (1 | day_of_week),
    latent_reporting_delay = c(0.3, 0.2, 0.1),
    data = pobs
  ),
  fit = enw_fit_opts(
    save_warmup = FALSE, pp = TRUE,
    chains = 2, iter_warmup = 500, iter_sampling = 500
  ),
  obs = enw_obs(family = "poisson", data = pobs)
)
```

The current example has some slight fitting complaints similar to those seen in #152 and is likely driven by bugs/non-optimal implementation choices.

Gulfa commented:
I wonder if there is an issue in the creation of the design matrices for the random walk in this PR. For the f-matrix we add a "c" to the variable name, ie if we have rw(week) we use cweekX as column names, but we do not add this "c" in the r-matrix. This means that the variable order can be different in the two matrices and we don't use the correct variances for the right betas. In my attempt at making a simple forecasting implementation, I think it worked to just remove the "c" prefix in the two functions used to generate the random walk design matrix. Possible that I have overlooked something important and that this is not an problem or that another solution is better.

adrian-lison commented on inst/examples/germany_latent_renewal.R:
```diff
@@ -0,0 +1,156 @@
+# - latent reporting delay representing the incubation period and assumed
+# ascertainment
+# - Day of week reporting effect
+expectation_module <- enw_expectation(
+  r = ~ rw(week),
+  generation_time = c(0.1, 0.4, 0.4, 0.1),
+  observation = ~ (1 | day_of_week),
+  latent_reporting_delay = 0.4 * c(0.05, 0.3, 0.6, 0.05),
+  data = pobs
+)
```
When testing this, I get a warning message

> Warning message:
In enw_add_metaobs_features(r_features, ...) :
  Pre-existing columns in `metaobs` will be overwritten: {day_of_week}.Pre-existing columns in `metaobs` will be overwritten: {day}.Pre-existing columns in `metaobs` will be overwritten: {week}.Pre-existing columns in `metaobs` will be overwritten: {month}.

seabbs replied:
This is from a recent PR from @pearsonca and isn't obviously desirable. We basically have to recreate the date helper variables as we have extended the date range of the data. In the current setup this throws a warning. We likely want to suppress this and ideally think of a more elegant way to handle this (though not clear what that is).

seabbs replied:
I've suppressed warnings from this function call for now.

adrian-lison reviewed (changes requested):
As always, very structured coding. I just have a few questions regarding the latent case modifier you have added and growth rate vs reproduction number aspects.

Also had some suggestions regarding naming and inline doc, will commit these in a second.

seabbs commented on R/model-modules.R:
```diff
@@ -219,76 +219,174 @@ enw_report <- function(non_parametric = ~0, structural = ~0, data) {
+#' renewal equation). When set to 1 (the default) this corresponds to modelling
+#' the daily growth rate.
+#'
+#' @param observation A formula (as implemented in [enw_formula()]) describing
+#' the modifiers used to adjust expected observations. This can use features
+#' defined by reference date as defined in `metareference` as produced by
+#' [enw_preprocess_data()]. By default no modifiers are used but a common choice
+#' might be to adjust for the day of the week. Note as the baseline is no
+#' modification an intercept is always used and it is set to 0.
```
As discussed at last weeks [community meeting](https://community.epinowcast.org/t/short-dated-epinowcast-meeting-2022-09-30/71/6)this really needs to be defined as a modifier to the probability of being reported (i.e as a discrete time hazard model). I think perhaps we should save doing this for another PR as this is again becoming quite long? What do you think?

adrian-lison replied:
Yep, sorry for all that spamming about the modiifer during my review, I was really misunderstanding things. Making another PR seems sensible

seabbs replied:
No worries. It wasn't exactly clear what was happening!

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -1,39 +1,74 @@
+  int expo_rncol;
+  matrix[expo_fnindex, expo_fncol + 1] expo_fdesign;
+  matrix[expo_fncol,  expo_rncol + 1] expo_rdesign;
+  array[2, 1] real expo_beta_sd_p;
```
No such thing as being overly pedantic! 

I think the docs are much improved with your changes.

Some of your concern in 1. I think stemmed from a slight misunderstanding about the role of the `observation` modifier (discussed [here](https://community.epinowcast.org/t/short-dated-epinowcast-meeting-2022-09-30/71/6)) but is still valid. I'm happy with your suggested change to the naming scheme but I am not sure it is as general as we might want (for example it won't always be cases). Do you still think it is a good idea to change and to change to this?

adrian-lison replied:
Exactly, I had misunderstood the goal of the modifier and totally agree that my proposed change does not make sense. Happy to either go back to "modifier" or to call it something like "latent_delay_effect".

seabbs replied:
latent_delay_modifier perhaps?

seabbs replied:
I am pretty lost with where we are on this. Do you remember? If yes could you make an issue so we can resolve ❤️

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -1,39 +1,74 @@
+  int expo_rncol;
+  matrix[expo_fnindex, expo_fncol + 1] expo_fdesign;
+  matrix[expo_fncol,  expo_rncol + 1] expo_rdesign;
+  array[2, 1] real expo_beta_sd_p;
```
As clarified in our [community meeting](https://community.epinowcast.org/t/short-dated-epinowcast-meeting-2022-09-30/71/6) this is a modifer for reporting practices such as day of the week effects etc. that should really be defined as a discrete time hazard model. As suggested in other responses I think we should handle this as a feature PR with related issue.

adrian-lison replied:
Agree!

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -139,29 +193,44 @@ transformed parameters{
+  if (expo_obs) {
+    expo_modifier = combine_effects(
+      0, expo_beta, expo_fdesign, expo_beta_sd, expo_rdesign, 1
+    );
```
As discussed in our [community meeting](https://community.epinowcast.org/t/short-dated-epinowcast-meeting-2022-09-30/71/6) we can identify this when it applies to reference day effects, such as weekly reporting practices where these aren't linked to the process driving infections. Also as discussed in that meeting this should really be defined as a discrete time to event model.

seabbs commented on inst/stan/functions/log_expected_obs_from_r.stan:
```diff
@@ -0,0 +1,21 @@
+          exp_lobs[k][r_seed + i] = local_r[i] +
+            log_sum_exp(
+              segment(exp_lobs[k], r_seed + i - gt_n, gt_n) + lrgt
+            );
```
Yes this is a good idea. Will add.

adrian-lison replied:
I think I already did

seabbs commented on R/epinowcast.R:
```diff
@@ -111,8 +111,10 @@ epinowcast <- function(data,
-                         formula = ~ rw(day, .group),
-                         order = 1,
+                         r = ~ rw(week, .group),
+                         generation_time = 1,
+                         observation = ~1,
+                         latent_reporting_delay = 1,
```
Good idea

seabbs commented on R/model-modules.R:
```diff
@@ -219,76 +219,174 @@ enw_report <- function(non_parametric = ~0, structural = ~0, data) {
-enw_expectation <- function(formula = ~ rw(day, .group), order = 1, data) {
-  if (as_string_formula(formula) %in% "~0") {
-    stop("An expectation model formula must be specified")
+enw_expectation <- function(r = ~ rw(day, by = .group), generation_time = 1,
+                            observation = ~1, latent_reporting_delay = 1,
```
Good idea

seabbs commented on inst/examples/germany_latent_renewal.R:
```diff
@@ -0,0 +1,156 @@
+expectation_module <- enw_expectation(
+  r = ~ rw(week),
+  generation_time = c(0.1, 0.4, 0.4, 0.1),
+  observation = ~ (1 | day_of_week),
```
No?

seabbs replied:
(This is a version of what is used in `EpiNow2` and is definitely identifable).

adrian-lison replied:
Sorry, I forgot you are enforcing a zero intercept there :D

seabbs commented on inst/stan/functions/log_expected_obs_from_r.stan:
```diff
@@ -0,0 +1,21 @@
+          exp_lobs[k][r_seed + i] = local_r[i] +
+            log_sum_exp(
+              segment(exp_lobs[k], r_seed + i - gt_n, gt_n) + lrgt
+            );
```
Yes I think we do need to test an alternative that is on the unit scale sadly :(

seabbs replied:
I haven't had a chance to get to this yet but hopefully will in the next few days.

seabbs replied:
I've played around with this in #168 and sadly it does seem to improve performance. I've made an issue (#173) to address so resolving here.

seabbs commented on inst/stan/functions/log_expected_obs_from_r.stan:
```diff
@@ -0,0 +1,21 @@
+          exp_lobs[k][r_seed + i] = local_r[i] +
+            log_sum_exp(
+              segment(exp_lobs[k], r_seed + i - gt_n, gt_n) + lrgt
+            );
```
So this is modelling the log of the instantaneous reproduction number which when the generation time is 1 will be the log approximation to the growth rate. Are you modelling Rt on the natural scale directly in your your work?

adrian-lison replied:
Ah okay, I was just confused here because everything is described as the growth rate and you used lower case `r` to denote this variable (and because the growth rate can of course be very different when we have an actual generation time distribution). Should we rename it to the reproduction number and use capital `R` instead then?

Nevertheless, we should decide whether the growth rate is also something we want to estimate/report alongside. If yes, it may make sense to model the growth rate and to compute R from it, because that is straightforward given a discrete generation time distribution, but not the other way round.

I have tried both: modeling Rt on the log scale and modeling Rt on the natural scale with a softplus link that avoids negative values.

seabbs replied:
I think we may want to report the daily growth rate but perhaps we can do that as post-processing and save ourselves some hassle in stan? 

In terms of what we model it has always seemed natural to use a log link to me to make modifiers multiplicative. I know that the Imperial group have made a lot of use of a logit link (to provide upper and lower bounds). That seems fine but does perhaps introduce some quite unexpected behaviour for uses. 

How have you found the softplus to perform? I am in principle open to it (in the future multiple links seem like an obvious nice feature to have). In terms of what we name it I really don't mind. Perhaps it should be called R and then be made explicit somewhere what the link we are using is?

seabbs replied:
Added an issue for looking at this (#176)

seabbs commented:
I rebased on `develop` to fix the many merge issues also caused by me.

adrian-lison commented on inst/examples/germany_latent_renewal.R:
```diff
@@ -0,0 +1,156 @@
+  r = ~ rw(week),
+  generation_time = c(0.1, 0.4, 0.4, 0.1),
+  observation = ~ (1 | day_of_week),
+  latent_reporting_delay = 0.4 * c(0.05, 0.3, 0.6, 0.05),
```
All cleared up as discussed. This is the actual scaling (in form of a workaround) that I confused the modifier with

seabbs commented:
# Summary

I started implementing natural scale stuff but thought I would just take a step back and look at some different model combinations.

Given the results below I think I am going to redirect from implementing log scale convolutions etc. to making this a minimal viable merge candidate. We can then look at optimisations in future PRs (obviously quite a bit of this performance is not ideal). I think that this needs:

- [ ] Making it possible for the growth rate model to have no intercept (this allows the default to stay the same as `main` - see experiment results which are encouraging).
- [ ] Model documentation
- [ ] Setting defaults 
- [ ] Checking documentation is clear about user options
- [ ] Changing the observation modifier to have a different name
- [ ] Adding a new argument for `latent_reporting_fraction` and enforcing `latent_reporting_delay` to sum to 1.
- [ ] Make issues for:
    - [ ] Exploring natural scale vs log scale renewal process
    - [ ] Exploring natural vs log scale latent reporting convolutions
    - [ ] Additional test coverage of the renewal infrastructure
    - [ ] Visualisations capabilities for the latent processes
    - [ ] Visualisations for the growth rate/reproduction number
    - [ ]Calculation of the daily growth rate (i.e regardless of generation time used)
    - [ ] Methodology issues for extending to time-varying and uncertain generation times and reporting distributions
   
# Experiments

I should have really done this in a big grid and made some plots etc but given this was a strike day I decided to work to contract and do a bad job (😆). Some runtimes using `inst/example/germany_latent_renewal.R` with various settings:

## Default

- All defaults: 330 seconds and 50+ DTs
- Random effect on Rt: 170 seconds
- Random effect on Rt with no intercept (requires manual stan code change for now): 220 seconds
- Weekly random walk on Rt: 200 seconds and 28 DTs
- Daily fixed effect on Rt: 320 seconds
- Fixed Rt: 100 seconds and no DTs
- No latent delays: 500+ (killed)
- No latent delays and a random effect on Rt: 150 seconds

## Growth rate

- All defaults: 435 seconds and 50+ DTs
- Random effect on growth rate: 245 seconds and 4 DTs 
- Random effect on the growth rate with no intercept (requires manual stan code change for now): 200 seconds
- Weekly random walk-on growth rate: 200 seconds and 28 DTs
- Daily fixed effect on growth rate: 456 seconds
- Fixed Rt: 100 seconds and no DTs
- No latent delays: 500+ (killed)
- No latent delays and a random effect on growth rate: 450 seconds
- No latent delays and a random effect on the growth rate with no intercept (requires manual stan code change for now): 200 seconds
- As above but for a maximum delay of 40 and with the settings used in the package readme: 80 seconds
- As above but with an intercept: 175 seconds

## Notes:

- Latent delays also include a day-of-the-week effect and this may be the cause of some of the speed-ups due to latent delays being present (as this data has such a clear day-of-the-week effect). 
- The random effect on growth rate with no intercept and no latent delays should be the model on `main`. This has comparable runtimes for me which is encouraging.

adrian-lison commented:
Thanks for the all the work @seabbs! I agree the random effect growth rate model without intercept could be a nice default. 

Mid-term, it might actually make sense to explore the stochastic version of the renewal process. I think this may really improve performance in a nowcasting setting: Whenever we are sampling a region of the parameter space where the reporting delays are not well calibrated (e.g. wrong weekday effects), we will have to cope with a potentially very irregular epidemic curve, and I could imagine that this can get the sampler stuck on a deterministic renewal process...

seabbs commented:
I am just fleshing out the issues for work left to do here (i.e noted in the review and in the list above) then plan to merge into `develop` so these bits can be fixed one by one vs in this monolithic PR. Happy to have push back on this!

seabbs reviewed (commented):
LGTM (self-review)
